internal/proxy/proxy.go

```diff
@@ -775,6 +767,58 @@ func newSocketMount(ctx context.Context, conf *Config, pc *portConfig, inst Inst
 	return m, nil
 }
 
+// newUnixSocketMount parses the configuration and returns the path to the unix
+// socket, or an error if that path is not valid.
+func newUnixSocketMount(inst InstanceConnConfig, unixSocketDir, version string) (string, error) {
+
+	var (
+		// the path to the unix socket
+		address string
+		// the parent directory of the unix socket
+		dir string
+	)
+
+	if inst.UnixSocketPath != "" {
+		// When UnixSocketPath is set
+		address = inst.UnixSocketPath
+
+		// If UnixSocketPath ends .s.PGSQL.5432, remove it for consistency
+		if strings.HasPrefix(version, "POSTGRES") && path.Base(address) == ".s.PGSQL.5432" {
+			address = path.Dir(address)
+		}
+
+		dir = path.Dir(address)
+	} else {
+		// When UnixSocket is set
+		dir = unixSocketDir
+		if dir == "" {
+			dir = inst.UnixSocket
+		}
+		address = UnixAddress(dir, inst.Name)
+	}
+
+	// if base directory does not exist, fail
+	if _, err := os.Stat(dir); err != nil {
+		return "", err
+	}
+
+	// When setting up a listener for Postgres, create address as a
+	// directory, and use the Postgres-specific socket name
+	// .s.PGSQL.5432.
+	if strings.HasPrefix(version, "POSTGRES") {
```

Comment:
Should we pull this into a function and re-use it above?

Reply:
I'm not sure it needs to be a function. I've refactored it into a parameter instead so that the check `strings.HasPrefix(version, "POSTGRES")` only appears once.